### PR TITLE
correct sending RST when request body fails

### LINKED ISSUE #38913: HTTP2: STREAM_CLOSED error after request cancellation
I found this issue by running @stephentoub's stress code from #38911 against local IIS, with the "GET w/ cancellation" operation enabled.

This results in a bunch of errors that look as follows:

```System.Net.Http.HttpRequestException: An error occurred while sending the request.
 ---> System.Net.Http.Http2ProtocolException: The HTTP/2 request failed with protocol error 'STREAM_CLOSED' (0x5).
   at System.Net.Http.Http2Connection.ProcessHeadersFrame(FrameHeader frameHeader) in C:\github\geoffkizer\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 316
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync() in C:\github\geoffkizer\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 224
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in C:\github\geoffkizer\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 1462
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) in C:\github\geoffkizer\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs:line 539
```

I believe what's happening here is:
(1) We are cancelling the request, sending a RST_STREAM, and removing the stream from the stream dictionary before we've received the entire response
(2) The server already has sent additional frames on this response before it processes the RST_STREAM
(3) We receive additional frames on this stream and treat this as an error

We need to change our processing of frames on closed streams to simply ignore them, not generate a stream error, as per #35652.

### LINKED ISSUE #39016: HTTP2: Cancel while sending request body causes RST_STREAM to be sent twice
Found as part of #38911 

If cancellation occurs while we are sending the request body, we will send RST_STREAM twice on the stream.

One occurs here: https://github.com/dotnet/corefx/blob/1833ec4769749dfce2b6dadad5882f208b499796/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs#L142

The other occurs here: https://github.com/dotnet/corefx/blob/1833ec4769749dfce2b6dadad5882f208b499796/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs#L1522

It seems like we only need one of these, but I'm not sure which is the correct one.

We need to ensure that the cancellation tests:
(a) validate that RST_STREAM is only sent once
(b) validate that nothing else is sent on the stream after RST_STREAM
(c) validate that the connection itself is still usable after the cancellation, i.e. a new request on the same connection succeeds.

@wfurt Can you take a look at this?

(edit: fix first code link)

### PR DESCRIPTION
**Pull request by wfurt**
When we cancel while sending RequestBody we could send multiple reset frames. This should be fixed now as well as I added new test to cover cases when cancel before and after body send starts. 

With new test, I verified that it fails before product fix. 
There was still some delay before second RST was sent so I added some delay to processing of second request to allow all tasks from first request to complete.  

fixes #38913 (test coverage) 
fixes #39016

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs [hunk 1]
```diff
@@ -1597,26 +1599,48 @@ public async Task Http2_PendingSend_SendsReset()
...
+                    frameCount++;
                  } while (frame.Type != FrameType.RstStream);
 
-                 frame = await connection.ReadFrameAsync(TimeSpan.FromMilliseconds(TestHelper.PassingTestTimeoutMilliseconds)).ConfigureAwait(false);
-                 Assert.Null(frame);    // Make sure we do not get any frames after getting RST.
+                 Assert.Equal(1, frame.StreamId);
+
+                frame = null;
+                (streamId, requestData) = await connection.ReadAndParseRequestHeaderAsync();
+                await connection.SendResponseHeadersAsync(streamId, endStream: false, HttpStatusCode.OK);
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs]: These 'await's don't have '.ConfigureAwait(false)'.  But later in this test, some of them do.  Why is this inconsistent?

Comment (wfurt) [reply src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http2.cs]: It should not be needed. Test it self is like application. I took some lines from Loopback server e.g. "library" and get extra .ConfigureAwait() from there.

**Comment by wfurt** [comment]
System.IO failures is unrelated.

**Pull request merged**